### DISCUSSION TITLE
Notes regarding replacing Splash

### DISCUSSION THREAD

**Discussion by Rafiot** [CIRCL/lookyloo#243] [Ideas]
## Why?
[Splash](https://github.com/scrapinghub/splash) is not really maintained anymore, and since the change from Scrapinghub to Zyte, it seems the opensource part is lacking and they're moving their services to a SaaS model. 
Relying on that isn't an option for Lookyloo, so it is time to start looking at alternatives.

Splash uses a pretty old Qt-WebKit thing, which means many new features in recent browser don't work.

## Requirements

* The capturing infrastructure must be easy. Preferably, something similar to the [aquarium setup](https://www.lookyloo.eu/docs/main/install-lookyloo-production.html)
* It must be able to capture multiple URLs in parallel, in separate environments.

* The capturing infrastructure must accept as input
  * referer
  * user agent
  * cookies
  * proxy
  * (optional) depth

* The capture must:
  * run in a fully featured browser in private mode, reset between each capture
  * wait for the page to be fully loaded
  * scroll down
  * wait a bit
  * (optional) allow other interactions with the page, clicks and so on

* The capture must return:
  * HAR file
  * Rendered content (HTML or other)
  * A screenshot of the whole page
  * The cookie jar
  * The last URL in the address bar (after redirect)

## Notes and references

* extract HAR from the dev tools in Firefox: https://defgsus.github.io/blog/2021/03/07/selenium-firefox-har-extract.html

# Playwright

## Initialize capture context 

* [Pass arbitrary headers](https://playwright.dev/python/docs/api/class-browsercontext#browser-context-set-extra-http-headers) -> referer, any other HTTP headers
* [User Agent](https://playwright.dev/python/docs/emulation#user-agent)
* [Viewport](https://playwright.dev/python/docs/emulation#viewport)
* [Pass cookies](https://playwright.dev/python/docs/api/class-browsercontext#browser-context-add-cookies)
* [Proxy](https://playwright.dev/python/docs/network#http-proxy)
* Depth: with instrumentation, not really used in prod right now anyway, subsequent queries are much better.

## Capture

* browser restarted between capture, no need for private mode
* [all kind of timers](https://playwright.dev/docs/actionability) to wait for all kind of things, this is the [easy one](https://playwright.dev/python/docs/api/class-page#page-goto)
* all the possible interactions

## Result
* [record_har_path](https://playwright.dev/python/docs/api/class-browser#browser-new-context) in browser context 
* [content](https://playwright.dev/python/docs/api/class-page#page-content)
* [screenshot](https://playwright.dev/python/docs/api/class-page#page-screenshot)
* [cookies](https://playwright.dev/python/docs/api/class-browsercontext#browser-context-cookies)
* [final URL](https://playwright.dev/python/docs/api/class-page#page-url)

And all that is available [in python](https://playwright.dev/python/docs/api/class-playwright)

**Comment by adulau** [comment]
# Potential strategies

- Using the [Headless Chromium](https://chromium.googlesource.com/chromium/src/+/lkgr/headless/README.md) and make a Python library to access it. 

# Good examples

- [chrome-php](https://github.com/chrome-php/chrome)

# Simple container/sandbox

- [firejail](https://github.com/netblue30/firejail/)

**Comment by esellier** [comment]
Alternative: 
Puppeteer (Headless Chromium) - dockerized or firejailed.
Use additional modules to fully mimic a true chromium : https://bot.sannysoft.com/

**Reply by Rafiot** [reply]
I had a look at Puppeteer some times ago but it wasn't able to get the HAR straight from the dev tools (there is a module that generates it programmatically, but it sounds like a terrible idea to me). 

Do you know of a way to do it via the dev tools in Puppeteer? `Dev tools > Network tab > right click > Save all as HAR with content` 
I cannot find anything like that in the docs.

**Reply by esellier** [reply]
Puppeteer allows to intercept all requests and responses.
It's quite easy to get all the required data to create the HAR.
But, yes, you have to do it manually, or rely on undocumented APIs to interact with the DevTools (which can also be done).

**Reply by Rafiot** [reply]
Is there any chance your Puppeteer script is open source by any chance? That would make it a lot easier without reinventing the wheel on my side.

**Reply by kimmoal** [reply]
I have a quick proof-of-concept API done in node (`express` with `puppeteer-cluster` doing the browser interaction). It is missing some splash features currently. Seems to take screenshots fine, HAR generation might have some issues with certain sites. Probably could run it alongside splash and see how much the output differs?

https://github.com/kimmoal/scrape-service/blob/master/scrape-service.js

can be built with `docker build -t scrape-service .`

Have not tried python puppeteer integration [`pyppeteer`](https://pypi.org/project/pyppeteer/) but it seems to be community effort compared to puppeteer which is somewhat a Google developed project so compatibility / long term support might me better with Chrome (or not, can't be sure with Google)

Other quite similar thing to puppeteer or splash is [Prerender](https://github.com/prerender/prerender) but this cannot generate content in the HAR file as it is using some built-in method which does not include response content, apparently. Also some other small things are missing, like cookies from response etc. If we are not interested in response content then this could work after some work.

Very ugly 5 minute Lookyloo integration in my branch that you can test quickly that it kind of works:

https://github.com/kimmoal/lookyloo/tree/puppeteer

**Reply by esellier** [reply]
Working on it... but I need to integrate HAR generation (instead of our own JSON format which contain less & more information).
I'm also considering using Microsoft Playwright (https://playwright.dev/) instead of Puppeteer. 
Playwright is cross-browser, has a nice API (giving more info, for requests timings for instance w/o coding CDP logic by yourself).

Regarding CDP, we found we could get ScriptParsed events (in addition to what we already do). Which is quite interesting in our case... As we always want to wait before ending the session (scraping task), so that we may catch new redirections initiated from JS.

**Reply by esellier** [reply]
Regarding puppeteer-cluster, be aware that sometimes the Chrome process won't close.
We modified the code so we can kill the Chrome process in that case.

**Reply by Rafiot** [reply]
Oh nice, thanks both of you for the pointers! 

@kimmoal I like puppeteer but as you said, it's google, which means they can decide to just drop support from one day to the next. 
The thing I love with the content of the chrome HAR export is the fact it contains the initiator, meaning the tree on Lookyloo would be much more precise, but it is a non-standard key, so it's not going to be compatible all the time and so on. 
I'm very interested by your branch, not sure I'll have time to play with it this month (kinda busy with other projects, sorry), but I'd love to see if we can get a drop-in replacement for splash with it. That would be pretty amazing (as you saw, the current code allows it without too much hassle).
Regarding Prerender, having the body in the HAR is quite important as we really want to be able to use the hashes to cross check against known files (legit and not). Loosing would be pretty annoying.

@esellier playwright look super interesting! I love the API, it looks solid, and seems to have HAR export with [recordHAR](https://playwright.dev/docs/api/class-browser#browser-new-context). And that gets us to an other thing I'd really love: as we get rid of splash, having an actual choice of browsers would be soo much nicer than just messing with user agents.

The main thing to keep in mind with lookyloo is that its priority is to allow an analyst to pinpoint interesting things happening on a page with a somewhat more readable view than the devtools interface. We have to support legitimate (news) websites that load 2000 URLs and download over 100MB of data in 50 seconds. I'm less interested in instrumenting the page (maybe with the exception of authorizing all the cookies).

**Reply by Rafiot** [reply]
okay, so I looked at the doc, and afaik, playwright seems to be doing all we need on lookyloo's side (see updated comments in the first message).

**Reply by esellier** [reply]
The goals we have:
- "stealthiness" : the scraping service must not (best effort as it's never perfect) be detected as a robot by targeted web sites
- "smart" : APIs and events which will allow to get all redirections (and especially (obfuscated and delayed) JS redirections)
- "thin & reliable" : no bug, no heavy resources consumption
- "LTS and supported" : based on a technology which will be supported on the long term, supported by the community
- "opened" : multi-platform, multi-browsers (inc. mobile), ...

Playwright meets these requirements:
- "stealthiness" : https://github.com/berstend/puppeteer-extra/issues/454 or we can port it on ourself as the Playwright API is quite similar to puppeteer's, and/or use Firefox instead of chromium to reinforce stealthiness
- "smart" : API (using CDP events and routes for intercepting scripts load requests) should be ok
- "thin & reliable": ... wait & see ...  missing the Cluster lib (but that is not a strong requirement)
- "LTS": let's believe Microsoft will support it
- "opened": OK

**Reply by Rafiot** [reply]
Sounds like we can share resources. I'll start poking around with playwright in the coming weeks and keep it a standalone python module.

**Reply by Rafiot** [reply]
Okay, so this script pretty much replaces splash on Lookyloo's side: https://github.com/Lookyloo/PlaywrightCapture
It's in a branch but will be merged and moved to prod in the next week or so.

**Reply by Rafiot** [reply]
Quick update: the playwright module is in place now, it's working much better than splash.

**Reply by esellier** [reply]
Works fine... !  
It passes my tests which failed with splash (delayed and obfuscated JS redirections).
Still missing some headers in requests so that it can pass some kind of filters.
Example of filtering URL which Lookyloo cannot pass: https://cpam-carte-vitalev3[.]com/
Other filtering URL: https://auth-ovhcloud-messervices[.]le-porte-voix[.]fr/clients/gKsxNb.php?verification#_

**Reply by esellier** [reply]
For the records, results for Bots filtering: https://lookyloo.circl.lu/tree/0e1be100-28af-4b2f-82f8-8c648f461ee9

**Reply by Rafiot** [reply]
Oh neat, I'm going to play with it and figure out how we can improve it. It's possible to pass arbitrary headers to the capture, so there is probably something to do about it.

**Reply by Rafiot** [reply]
Okay, so I looked at it a bit, and the bot detection result is a lot better if the browser isn't started as headless. That's not really an option on a server, but I leave it here as a datapoint.

One module that might be interesting to use is this one: https://github.com/AtuboDad/playwright_stealth I haven't yet, but it might be useful.

**Reply by esellier** [reply]
Running the ("headed") browser,  not headless, can be done with a pseudo
X-Server:
https://www.x.org/archive/X11R7.6/doc/man/man1/Xvfb.1.xhtml
Not tested though...

This module seems to be the transposition of JS code mentioned here:
[berstend/puppeteer-extra#454 (comment)](https://github.com/berstend/puppeteer-extra/issues/454#issuecomment-917437212)

It works quite well with Firefox also (as it does not require a CDP
session).

**Comment by Rafiot** [comment]
Few updates on that front.

* BrowserMob Proxy (selenium thing) can export a HAR, but the HAR is still generated by the proxy itself instead of by exporting it straight from the dev tools. And on top of that, the HAR generating code hasn't been touched for 5 years or so.
* It is possible to get selenium to trigger commands inside the firefox devtools from a [headless firefox browser](https://stackoverflow.com/questions/63699473/is-the-firefox-web-console-accessible-in-headless-mode/63708393#63708393) which means we should be able to simulate a "save as HAR" using something [similar to that](https://github.com/firefox-devtools/har-export-trigger/blob/master/lib/harapi.js)
* The moilla dev docs seems to have comething very helpful: https://developer.mozilla.org/en-US/docs/Mozilla/Add-ons/WebExtensions/API/devtools/network/getHAR

**Reply by kimmoal** [reply]
getHAR does not return content as far as I know, are we interested in getting that?

**Reply by Rafiot** [reply]
In order to build the tree on Lookyloo, we must get a HAR file one way or the other (that's what [har2tree](https://github.com/Lookyloo/har2tree) requires). 

My understanding is that getHAR only works from an extension (?), but I'm not totally sure and never played with it.
